File: Nudge-Neurodivergent-task-tracker/social_support.py
"""
Social support features for accountability and body doubling
"""

from datetime import datetime, timedelta
from enum import Enum
from typing import List, Optional, Dict
from dataclasses import dataclass


class SessionType(Enum):
    """Types of social support sessions"""
    BODY_DOUBLING = "Body Doubling"
    ACCOUNTABILITY_CHECKIN = "Accountability Check-in"
    CELEBRATION = "Achievement Celebration"
    SUPPORT_CALL = "Support Call"


@dataclass
class BodyDoublingSession:
    """A body doubling session for parallel work"""
    name: str
    start_time: datetime
    duration_minutes: int
    task_focus: Optional[str] = None
    session_id: str = None
    ended: bool = False
    
    def __post_init__(self):
        if self.session_id is None:
            self.session_id = f"bd_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
    
    def get_end_time(self) -> datetime:
        """Get session end time"""
        return self.start_time + timedelta(minutes=self.duration_minutes)
    
    def is_active(self) -> bool:
        """Check if session is currently active"""
        return not self.ended and datetime.now() < self.get_end_time()
    
    def get_time_remaining(self) -> str:
        """Get remaining time in session"""
        if self.ended:
            return "Session ended"
        
        remaining = self.get_end_time() - datetime.now()
        if remaining.total_seconds() <= 0:
            return "Time's up!"
        
        minutes = int(remaining.total_seconds() // 60)
        seconds = int(remaining.total_seconds() % 60)
        return f"{minutes:02d}:{seconds:02d}"
    
    def end_session(self):
        """End the session early"""
        self.ended = True


@dataclass
class AccountabilityPartner:
    """An accountability partner or support person"""
    name: str
    contact_method: str  # "text", "call", "email", etc.
    preferred_checkin_frequency: str  # "daily", "weekly", etc.
    last_checkin: Optional[datetime] = None
    shared_goals: List[str] = None
    
    def __post_init__(self):
        if self.shared_goals is None:
            self.shared_goals = []
    
    def is_checkin_due(self) -> bool:
        """Check if a check-in is due"""
        if self.last_checkin is None:
            return True
        
        days_since = (datetime.now() - self.last_checkin).days
        
        if self.preferred_checkin_frequency == "daily":
            return days_since >= 1
        elif self.preferred_checkin_frequency == "weekly":
            return days_since >= 7
        elif self.preferred_checkin_frequency == "biweekly":
            return days_since >= 14
        
        return False


@dataclass
class Achievement:
    """An achievement to share or celebrate"""
    title: str
    description: str
    date_achieved: datetime
    category: str  # "task", "habit", "streak", "milestone"
    shared_with: List[str] = None
    
    def __post_init__(self):
        if self.shared_with is None:
            self.shared_with = []


class SocialSupport:
    """Manages social support features"""
    
    def __init__(self):
        self.accountability_partners: List[AccountabilityPartner] = []
        self.active_session: Optional[BodyDoublingSession] = None
        self.session_history: List[BodyDoublingSession] = []
        self.achievements: List[Achievement] = []
        self._create_sample_data()
    
    def _create_sample_data(self):
        """Create sample accountability partners"""
        sample_partner = AccountabilityPartner(
            name="Study Buddy",
            contact_method="text",
            preferred_checkin_frequency="weekly",
            shared_goals=["Task completion", "Daily habits"]
        )
        self.accountability_partners.append(sample_partner)
    
    def start_body_doubling_session(self, name: str, duration_minutes: int, 
                                  task_focus: Optional[str] = None) -> BodyDoublingSession:
        """Start a new body doubling session"""
        if self.active_session and self.active_session.is_active():
            print("🚨 Another session is already active!")
            return self.active_session
        
        session = BodyDoublingSession(
            name=name,
            start_time=datetime.now(),
            duration_minutes=duration_minutes,
            task_focus=task_focus
        )
        
        self.active_session = session
        print(f"👥 Body doubling session started: {name}")
        print(f"⏰ Duration: {duration_minutes} minutes")
        if task_focus:
            print(f"🎯 Focus: {task_focus}")
        print("💡 Work alongside others virtually - you're not alone!")
        
        return session
    
    def end_current_session(self):
        """End the current body doubling session"""
        if not self.active_session or not self.active_session.is_active():
            print("🌸 No active session to end")
            return
        
        self.active_session.end_session()
        self.session_history.append(self.active_session)
        
        duration = (datetime.now() - self.active_session.start_time).total_seconds() / 60
        print(f"✅ Session completed! You focused for {duration:.0f} minutes")
        print("🎉 Great job working together!")
        
        self.active_session = None
    
    def get_current_session(self) -> Optional[BodyDoublingSession]:
        """Get current active session"""
        if self.active_session and self.active_session.is_active():
            return self.active_session
        return None
    
    def add_accountability_partner(self, name: str, contact_method: str, 
                                 frequency: str, shared_goals: List[str] = None):
        """Add a new accountability partner"""
        partner = AccountabilityPartner(
            name=name,
            contact_method=contact_method,
            preferred_checkin_frequency=frequency,
            shared_goals=shared_goals or []
        )
        self.accountability_partners.append(partner)
        print(f"🤝 Added accountability partner: {name}")
    
    def record_checkin(self, partner_name: str):
        """Record a check-in with an accountability partner"""
        for partner in self.accountability_partners:
            if partner.name.lower() == partner_name.lower():
                partner.last_checkin = datetime.now()
                print(f"📞 Check-in recorded with {partner.name}!")
                return True
        return False
    
    def get_due_checkins(self) -> List[str]:
        """Get list of overdue check-ins"""
        due_checkins = []
        for partner in self.accountability_partners:
            if partner.is_checkin_due():
                due_checkins.append(f"Check in with {partner.name} ({partner.contact_method})")
        return due_checkins
    
    def add_achievement(self, title: str, description: str, category: str):
        """Add an achievement to celebrate"""
        achievement = Achievement(
            title=title,
            description=description,
            date_achieved=datetime.now(),
            category=category
        )
        self.achievements.append(achievement)
        print(f"🏆 Achievement unlocked: {title}!")
        print(f"   {description}")
    
    def get_celebration_reminders(self) -> List[str]:
        """Get reminders to celebrate recent achievements"""
        recent_achievements = [
            a for a in self.achievements
            if (datetime.now() - a.date_achieved).days <= 7
        ]
        
        if recent_achievements:
            return [f"🎉 Celebrate: {a.title}" for a in recent_achievements[-2:]]
        return []
    
    def display_social_dashboard(self):
        """Display social support dashboard"""
        print("\n👥 SOCIAL SUPPORT DASHBOARD")
        print("=" * 50)
        
        # Active session
        if self.active_session and self.active_session.is_active():
            print(f"🔴 ACTIVE SESSION: {self.active_session.name}")
            print(f"   Time remaining: {self.active_session.get_time_remaining()}")
            if self.active_session.task_focus:
                print(f"   Focus: {self.active_session.task_focus}")
        else:
            print("⚪ No active body doubling session")
        
        # Accountability partners
        print(f"\n🤝 ACCOUNTABILITY PARTNERS ({len(self.accountability_partners)}):")
        for partner in self.accountability_partners:
            status = "✅" if not partner.is_checkin_due() else "🟡"
            last_checkin = partner.last_checkin.strftime("%Y-%m-%d") if partner.last_checkin else "Never"
            print(f"   {status} {partner.name} | Last: {last_checkin} | {partner.contact_method}")
        
        # Recent achievements
        recent_achievements = [
            a for a in self.achievements
            if (datetime.now() - a.date_achieved).days <= 30
        ]
        
        if recent_achievements:
            print(f"\n🏆 RECENT ACHIEVEMENTS ({len(recent_achievements)}):")
            for achievement in recent_achievements[-3:]:
                days_ago = (datetime.now() - achievement.date_achieved).days
                print(f"   🌟 {achievement.title} ({days_ago} days ago)")
        
        # Session stats
        total_sessions = len(self.session_history)
        total_time = sum((s.get_end_time() - s.start_time).total_seconds() / 60 
                        for s in self.session_history if s.ended)
        
        print(f"\n📊 STATS:")
        print(f"   Total sessions: {total_sessions}")
        print(f"   Total focus time: {total_time:.0f} minutes ({total_time/60:.1f} hours)")
        
        print("=" * 50)
    
    def get_body_doubling_tips(self) -> List[str]:
        """Get tips for effective body doubling"""
        return [
            "🎯 Set a specific task or goal for the session",
            "📱 Agree on minimal talking to maintain focus",
            "⏰ Use shared timers for breaks and check-ins",
            "🌸 Be supportive - presence is more important than productivity",
            "💻 Consider virtual background if camera makes you uncomfortable",
            "☕ Schedule short breaks together",
            "🎵 Agree on background music/sounds preferences",
            "📝 Share quick wins and struggles during breaks",
            "🚀 Celebrate completed sessions together",
            "💛 Remember: parallel presence reduces isolation"
        ]

File: Nudge-Neurodivergent-task-tracker/SocialSupport.java
import java.time.LocalDateTime;
import java.time.Duration;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;

/**
 * Manages body doubling sessions, accountability, and social support features
 */
public class SocialSupport {
    public enum SessionType {
        BODY_DOUBLING("👥 Body Doubling"),
        ACCOUNTABILITY("📞 Accountability Check-in"),
        CELEBRATION("🎉 Achievement Celebration"),
        SUPPORT("🤗 Support Session");
        
        private final String displayName;
        
        SessionType(String displayName) {
            this.displayName = displayName;
        }
        
        public String getDisplayName() { return displayName; }
    }
    
    public static class BodyDoublingSession {
        private LocalDateTime startTime;
        private int plannedDurationMinutes;
        private SessionType type;
        private String description;
        private Task associatedTask;
        private boolean isActive;
        private LocalDateTime endTime;
        
        public BodyDoublingSession(SessionType type, int durationMinutes, String description, Task task) {
            this.startTime = LocalDateTime.now();
            this.plannedDurationMinutes = durationMinutes;
            this.type = type;
            this.description = description;
            this.associatedTask = task;
            this.isActive = true;
        }
        
        public void endSession() {
            this.isActive = false;
            this.endTime = LocalDateTime.now();
        }
        
        public int getActualDurationMinutes() {
            LocalDateTime end = isActive ? LocalDateTime.now() : endTime;
            return (int) Duration.between(startTime, end).toMinutes();
        }
        
        public String getTimeRemaining() {
            if (!isActive) return "Session ended";
            
            int elapsed = getActualDurationMinutes();
            int remaining = Math.max(0, plannedDurationMinutes - elapsed);
            return remaining + " minutes remaining";
        }
        
        // Getters
        public LocalDateTime getStartTime() { return startTime; }
        public int getPlannedDurationMinutes() { return plannedDurationMinutes; }
        public SessionType getType() { return type; }
        public String getDescription() { return description; }
        public Task getAssociatedTask() { return associatedTask; }
        public boolean isActive() { return isActive; }
        public LocalDateTime getEndTime() { return endTime; }
    }
    
    public static class AccountabilityPartner {
        private String name;
        private String relationshipType; // friend, family, colleague, etc.
        private List<String> checkInSchedule; // days of week
        private LocalDateTime lastCheckIn;
        private int totalCheckIns;
        private boolean isActive;
        
        public AccountabilityPartner(String name, String relationshipType) {
            this.name = name;
            this.relationshipType = relationshipType;
            this.checkInSchedule = new ArrayList<>();
            this.totalCheckIns = 0;
            this.isActive = true;
        }
        
        public void recordCheckIn() {
            this.lastCheckIn = LocalDateTime.now();
            this.totalCheckIns++;
        }
        
        public boolean isCheckInDue() {
            if (lastCheckIn == null) return true;
            
            Duration timeSinceCheck = Duration.between(lastCheckIn, LocalDateTime.now());
            return timeSinceCheck.toDays() >= 7; // Weekly check-ins
        }
        
        // Getters and setters
        public String getName() { return name; }
        public String getRelationshipType() { return relationshipType; }
        public List<String> getCheckInSchedule() { return new ArrayList<>(checkInSchedule); }
        public LocalDateTime getLastCheckIn() { return lastCheckIn; }
        public int getTotalCheckIns() { return totalCheckIns; }
        public boolean isActive() { return isActive; }
        
        public void setActive(boolean active) { this.isActive = active; }
        public void addCheckInDay(String day) { checkInSchedule.add(day); }
        public void removeCheckInDay(String day) { checkInSchedule.remove(day); }
    }
    
    public static class Achievement {
        private String title;
        private String description;
        private LocalDateTime achievedDate;
        private boolean isShared;
        
        public Achievement(String title, String description) {
            this.title = title;
            this.description = description;
            this.achievedDate = LocalDateTime.now();
            this.isShared = false;
        }
        
        public void markAsShared() { this.isShared = true; }
        
        // Getters
        public String getTitle() { return title; }
        public String getDescription() { return description; }
        public LocalDateTime getAchievedDate() { return achievedDate; }
        public boolean isShared() { return isShared; }
        
        @Override
        public String toString() {
            String shareStatus = isShared ? "📢 Shared" : "🤐 Private";
            return String.format("🏆 %s - %s (%s)", title, description, shareStatus);
        }
    }
    
    private List<BodyDoublingSession> sessions;
    private List<AccountabilityPartner> partners;
    private List<Achievement> achievements;
    private BodyDoublingSession currentSession;
    
    public SocialSupport() {
        this.sessions = new ArrayList<>();
        this.partners = new ArrayList<>();
        this.achievements = new ArrayList<>();
    }
    
    public BodyDoublingSession startBodyDoublingSession(int durationMinutes, String description, Task task) {
        if (currentSession != null && currentSession.isActive()) {
            currentSession.endSession();
        }
        
        currentSession = new BodyDoublingSession(SessionType.BODY_DOUBLING, durationMinutes, description, task);
        sessions.add(currentSession);
        
        System.out.println("\n👥 BODY DOUBLING SESSION STARTED");
        System.out.println("=".repeat(40));
        System.out.println("⏰ Duration: " + durationMinutes + " minutes");
        System.out.println("📋 Focus: " + description);
        if (task != null) {
            System.out.println("🎯 Task: " + task.getName());
        }
        System.out.println("\n💡 Body Doubling Tips:");
        System.out.println("   • You're not alone - others are working too");
        System.out.println("   • Focus on your own work, don't compare");
        System.out.println("   • Take breaks when you need them");
        System.out.println("   • Celebrate small progress");
        System.out.println("=".repeat(40));
        
        return currentSession;
    }
    
    public void endCurrentSession() {
        if (currentSession != null && currentSession.isActive()) {
            currentSession.endSession();
            
            System.out.println("\n🎉 BODY DOUBLING SESSION COMPLETE!");
            System.out.println("⏰ Actual duration: " + currentSession.getActualDurationMinutes() + " minutes");
            System.out.println("✨ " + getSessionCompletionMessage());
            
            currentSession = null;
        }
    }
    
    public void addAccountabilityPartner(String name, String relationshipType) {
        partners.add(new AccountabilityPartner(name, relationshipType));
        System.out.println("🤝 Added " + name + " as accountability partner!");
    }
    
    public void recordCheckIn(String partnerName) {
        AccountabilityPartner partner = findPartner(partnerName);
        if (partner != null) {
            partner.recordCheckIn();
            System.out.println("📞 Check-in recorded with " + partnerName + "!");
            System.out.println("💬 " + getCheckInPrompt());
        }
    }
    
    private AccountabilityPartner findPartner(String name) {
        return partners.stream()
            .filter(p -> p.getName().equalsIgnoreCase(name))
            .findFirst()
            .orElse(null);
    }
    
    public void addAchievement(String title, String description) {
        achievements.add(new Achievement(title, description));
        System.out.println("\n🏆 NEW ACHIEVEMENT UNLOCKED!");
        System.out.println("✨ " + title + " - " + description);
        System.out.println("🎉 " + getCelebrationMessage());
    }
    
    public void shareAchievement(String title) {
        Achievement achievement = achievements.stream()
            .filter(a -> a.getTitle().equalsIgnoreCase(title))
            .findFirst()
            .orElse(null);
        
        if (achievement != null) {
            achievement.markAsShared();
            System.out.println("📢 Achievement shared: " + title);
            System.out.println("💪 Sharing your wins helps inspire others!");
        }
    }
    
    public List<String> getBodyDoublingPrompts() {
        return List.of(
            "👥 Working alongside others, even virtually",
            "🎯 Stay focused on your own task",
            "💡 Others are tackling their challenges too",
            "⏰ Take breaks when your brain needs them",
            "🌟 Your presence helps others stay motivated",
            "🤝 You're part of a supportive community",
            "📈 Every minute of focus is progress",
            "💪 You're building great work habits"
        );
    }
    
    public List<String> getDueCheckIns() {
        List<String> dueCheckIns = new ArrayList<>();
        
        for (AccountabilityPartner partner : partners) {
            if (partner.isActive() && partner.isCheckInDue()) {
                dueCheckIns.add("📞 Check-in due with " + partner.getName());
            }
        }
        
        return dueCheckIns;
    }
    
    public void displaySocialDashboard() {
        System.out.println("\n" + "=".repeat(50));
        System.out.println("🤝 SOCIAL SUPPORT DASHBOARD");
        System.out.println("=".repeat(50));
        
        // Current session
        if (currentSession != null && currentSession.isActive()) {
            System.out.println("🟢 ACTIVE SESSION:");
            System.out.println("   " + currentSession.getType().getDisplayName());
            System.out.println("   " + currentSession.getTimeRemaining());
            if (currentSession.getAssociatedTask() != null) {
                System.out.println("   Working on: " + currentSession.getAssociatedTask().getName());
            }
        } else {
            System.out.println("⚪ No active session");
        }
        
        // Accountability partners
        System.out.println("\n👥 ACCOUNTABILITY PARTNERS:");
        if (partners.isEmpty()) {
            System.out.println("   💡 Consider adding someone for regular check-ins!");
        } else {
            for (AccountabilityPartner partner : partners) {
                if (partner.isActive()) {
                    String status = partner.isCheckInDue() ? "📞 Due" : "✅ Current";
                    System.out.printf("   %s %s (%s) - %d check-ins\n", 
                        status, partner.getName(), partner.getRelationshipType(), partner.getTotalCheckIns());
                }
            }
        }
        
        // Recent achievements
        System.out.println("\n🏆 RECENT ACHIEVEMENTS:");
        List<Achievement> recentAchievements = achievements.stream()
            .sorted((a, b) -> b.getAchievedDate().compareTo(a.getAchievedDate()))
            .limit(3)
            .toList();
        
        if (recentAchievements.isEmpty()) {
            System.out.println("   🌟 Your achievements will appear here!");
        } else {
            for (Achievement achievement : recentAchievements) {
                System.out.println("   " + achievement);
            }
        }
        
        // Session stats
        System.out.println("\n📊 SESSION STATS:");
        int totalSessions = sessions.size();
        int totalMinutes = sessions.stream().mapToInt(BodyDoublingSession::getActualDurationMinutes).sum();
        System.out.printf("   Total sessions: %d\n", totalSessions);
        System.out.printf("   Total focus time: %d minutes (%.1f hours)\n", totalMinutes, totalMinutes / 60.0);
        
        System.out.println("=".repeat(50));
    }
    
    private String getSessionCompletionMessage() {
        String[] messages = {
            "You showed up and did the work - that's what matters!",
            "Look at you building that focus muscle!",
            "Every session makes the next one easier!",
            "You're proving to yourself that you can do this!",
            "That's how we build sustainable work habits!",
            "Your brain is learning to focus better each time!"
        };
        
        return messages[(int) (Math.random() * messages.length)];
    }
    
    private String getCheckInPrompt() {
        String[] prompts = {
            "What's one win you want to share?",
            "What's been challenging lately?",
            "What support do you need this week?",
            "What are you proud of recently?",
            "What's your focus for the next few days?",
            "How are you taking care of yourself?"
        };
        
        return prompts[(int) (Math.random() * prompts.length)];
    }
    
    private String getCelebrationMessage() {
        String[] messages = {
            "This is exactly the kind of progress we love to see!",
            "You're building something amazing, one step at a time!",
            "Look at you showing up for yourself!",
            "This achievement represents real growth!",
            "You should feel proud of this accomplishment!",
            "Every achievement builds momentum for the next one!"
        };
        
        return messages[(int) (Math.random() * messages.length)];
    }
    
    public List<String> getSocialMotivationTips() {
        return List.of(
            "🤝 Connection reduces the isolation that often comes with ADHD",
            "👥 Body doubling works because it provides gentle accountability",
            "🗣️ Sharing struggles makes them feel less overwhelming",
            "🎉 Celebrating wins together amplifies the positive feelings",
            "💪 Others believe in you even when you don't believe in yourself",
            "🌟 Your success inspires others in their journey too",
            "🤗 It's okay to need support - that's what community is for"
        );
    }
    
    // Getters
    public List<BodyDoublingSession> getAllSessions() { return new ArrayList<>(sessions); }
    public List<AccountabilityPartner> getPartners() { return new ArrayList<>(partners); }
    public List<Achievement> getAchievements() { return new ArrayList<>(achievements); }
    public BodyDoublingSession getCurrentSession() { return currentSession; }
}